appl/test/auditdns.c

```diff
@@ -65,32 +68,308 @@ gethostbyname2(const char *name, int af)
-getaddrinfo_fn_t getaddrinfo;
+void
+freeaddrinfo(struct addrinfo *ai)
+{
+
+    free(ai->ai_addr);
+    free(ai);
+}
+
 int
 getaddrinfo(const char *hostname, const char *servname,
     const struct addrinfo *restrict hints,
     struct addrinfo **restrict res)
 {
-    void *sym;
+    char *servend;
+    unsigned long port;
+    union {
+	struct sockaddr		sa;
+	struct sockaddr_in	sin;
+	struct sockaddr_in6	sin6;
+    } *addr = NULL;
+    int af[2] = {AF_INET, AF_INET6};
+    socklen_t addrlen[2] = {sizeof(addr->sin), sizeof(addr->sin6)};
+    int socktype[2] = {SOCK_DGRAM, SOCK_STREAM};
+    int proto[2] = {IPPROTO_UDP, IPPROTO_TCP};
+    size_t i, j, naddr, nproto;
+    struct addrinfo *ai = NULL;
+    int error;
 
+    /*
+     * DNS audit: Abort unless the user specified hints with
+     * AI_NUMERICHOST, AI_NUMERICSERV, and no AI_CANONNAME.
+     */
     if (hints == NULL ||
 	(hints->ai_flags & AI_NUMERICHOST) == 0 ||
+	(hints->ai_flags & AI_NUMERICSERV) == 0 ||
 	(hints->ai_flags & AI_CANONNAME) != 0) {
 	fprintf(stderr, "DNS leak: %s %s:%s\n",
 	    __func__, hostname, servname);
 	abort();
     }
 
-    if ((sym = dlsym(RTLD_NEXT, __func__)) == NULL) {
-	fprintf(stderr, "dlsym(RTLD_NEXT, \"%s\") failed: %s\n",
-	    __func__, dlerror());
-	return EAI_FAIL;
+    /*
+     * Check hints for address family.  If unspecified, use the default
+     * set of address families: {AF_INET, AF_INET6}.
+     */
+    switch (hints->ai_family) {
+    case AF_UNSPEC:
+	naddr = 2;
+	break;
+    case AF_INET:
+	naddr = 1;
+	af[0] = AF_INET;
+	addrlen[0] = sizeof(addr->sin);
+	break;
+    case AF_INET6:
+	naddr = 1;
+	af[0] = AF_INET6;
+	addrlen[0] = sizeof(addr->sin6);
+	break;
+    default:
+	error = EAI_FAMILY;
+	goto out;
+    }
+
+    /*
+     * Check hints for socket type and protocol.  If both are zero, we
+     * use the default set of socktype/proto pairs.  If one is
+     * specified but not the other, use the default.  If both are
+     * specified, make sure they match.
+     */
+    switch (hints->ai_socktype) {
+    case 0:
+	if (hints->ai_protocol == 0)
+	    nproto = sizeof(proto)/sizeof(proto[0]);
+	else
+	    nproto = 1;
+	break;
+    case SOCK_DGRAM:		/* datagram <-> UDP */
+	if (hints->ai_protocol != 0 && hints->ai_protocol != IPPROTO_UDP) {
+	    error = EAI_SOCKTYPE;;
+	    goto out;
+	}
+	socktype[0] = SOCK_DGRAM;
+	proto[0] = IPPROTO_UDP;
+	nproto = 1;
+	break;
+    case SOCK_STREAM:		/* stream <-> TCP */
+	if (hints->ai_protocol != 0 && hints->ai_protocol != IPPROTO_TCP) {
+	    error = EAI_SOCKTYPE;
+	    goto out;
+	}
+	socktype[0] = SOCK_STREAM;
+	proto[0] = IPPROTO_TCP;
+	nproto = 1;
+	break;
+    default:
+	error = EAI_SOCKTYPE;
+	goto out;
+    }
+
+    /*
+     * Check whether a service is specified at all.
+     */
+    if (servname == NULL) {
+	/*
+	 * No service specified.  Use the wildcard port 0.
+	 */
+	port = 0;
+    } else {
+	/*
+	 * Service specified.  Parse it as a nonnegative integer, at
```

Comment:
One might think that `strtoul()` would refuse to parse negative numbers, but, no, it doesn't.  What I do for this sort of thing is use `strtol()`/`strtoll()`/`strtoimax()` as appropriate, check that the the result is valid, not negative, then use the result, and if I have to support numeric values larger than the signed result type's largest integer then I explicitly look for a `-` sign.  What a pain.  Not saying you have to here -- this is a test program and there will not be negative numbers here.  I'm just... ranting.

Reply:
No need.  Negative numbers will be rejected anyways just because they would be larger than 65535 anyways.  I merely felt the need to whine about `strtoul()` :)